cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaNodeSetController.java

```diff
@@ -0,0 +1,478 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.assembly;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.fabric8.kubernetes.api.model.DeletionPropagation;
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.OwnerReference;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.fabric8.kubernetes.client.informers.cache.Lister;
+import io.fabric8.kubernetes.client.internal.readiness.Readiness;
+import io.strimzi.api.kafka.KafkaList;
+import io.strimzi.api.kafka.KafkaNodeSetList;
+import io.strimzi.api.kafka.model.Kafka;
+import io.strimzi.api.kafka.model.KafkaNodeSet;
+import io.strimzi.api.kafka.model.KafkaNodeSetBuilder;
+import io.strimzi.api.kafka.model.status.KafkaNodeSetStatus;
+import io.strimzi.operator.cluster.model.ModelUtils;
+import io.strimzi.operator.cluster.model.StatusDiff;
+import io.strimzi.operator.cluster.operator.resource.PodRevision;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.operator.resource.CrdOperator;
+import io.strimzi.operator.common.operator.resource.PodOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.stream.Collectors;
+
+/**
+ * KafkaNodeSet controller is responsible for managing the KafkaNodeSets and the pods which belong to them
+ */
+public class KafkaNodeSetController implements Runnable {
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaNodeSetController.class);
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private boolean stop = false;
+
+    private final PodOperator podOperator;
+    private final CrdOperator<KubernetesClient, KafkaNodeSet, KafkaNodeSetList> kafkaNodeSetOperator;
+    private final Optional<LabelSelector> crSelector;
+    private final String watchedNamespace;
+
+    private final BlockingQueue<SimplifiedReconciliation> workQueue = new ArrayBlockingQueue<>(1024);
```

Comment:
What are the advantages of using our own work queue and worker thread over using a single threaded `Executor`?

Reply:
I didn't really consider `Executor` or `ExecutorService`. I follow this is a more common controller pattern. The only advantage I see would be if we use a thread pool to parallelize this. But that brings additional complexity around locking etc. So I'm not sure I see any benefits in that. Do you see any benefits in using Executor / ExecutorService?

Reply:
Nothing completely convincing:

1. Not having to think so hard about threading issues (e.g. around stopping the worker thread).
2. Built-in support for cancellation (right now it doesn't look like we need that).
3. An abstraction which could be useful (e.g. for testing purposes).